### PR TITLE
[gitlab] Allow docker release jobs to target dev repositories on non-tagged pipelines

### PR DESCRIPTION
**Pull request by KSerrania**
### What does this PR do?

Allows docker release jobs to release images to the `agent-dev` repo on non-tagged deploy pipelines.

Refactors the release jobs: uses a common script for Dockerhub and GCR jobs, splits the dogstatsd release jobs.

First part of a 2-part refactor of the Gitlab jobs used to release the docker images. The second part will move these jobs to dedicated child pipelines.

### Motivation

Be able to test the docker release jobs outside of releases, without releasing images to the prod repositories.

### Additional Notes

Currently, in each job, there is a bash / powershell line which adds `-dev` to the repository name if we don't follow the `deploy_on_tag_*` rules (so that only `deploy_on_tag_*` pipelines can release to prod repositories). This is a temporary workaround: the long-term solution would be to use Gitlab rules directly to set variables, but that feature got added in Gitlab 13.8, which we don't use yet.

The release documentation will be updated once the 2nd part of the refactor is done.

### Testing plan

Run all docker release jobs on a non-tagged pipeline, check that they succeed & that they release images to the `-dev` repositories.

### CODE REVIEW

**Review by vboulineau — APPROVED** [review]
LGTM, we'll simplify this in the future with the introduction of a dedicated pipeline for public images.

**Review by albertvaka — COMMENTED** [review]
This change will make us push the `latest` tag on `agent-dev` which previously was never tagged. There might be the expectation that `agent-dev:latest` points to the latest agent build, but this might not be the case since the job that regularly pushes to `agent-dev` won't update latest (only the manual jobs, when we trigger them). To avoid confusion, we might want to update the job that deploys to `agent-dev` so they also tag `latest` on every build.

**Review by albertvaka — APPROVED** [review]

File: .gitlab/docker_common/tag_job_templates.yml [hunk 1]
```diff
@@ -93,6 +93,7 @@
   variables:
     <<: *google_container_registry_variables
   before_script:
+    - '$_instance_id = (iwr  -UseBasicParsing http://169.254.169.254/latest/meta-data/instance-id).content ; Write-Host "Running on instance $($_instance_id)"'
```
Comment (albertvaka) [inline .gitlab/docker_common/tag_job_templates.yml:96]:
```suggestion
    - $_instance_id = (iwr  -UseBasicParsing http://169.254.169.254/latest/meta-data/instance-id).content
    - Write-Host "Running on instance $($_instance_id)"
```

Comment (KSerrania) [reply .gitlab/docker_common/tag_job_templates.yml:96]: If you want to do that, we should do it in a separate PR. This line appears a lot of times across all our gitlab files 😄

**Review by celenechang — APPROVED** [review]

**Comment by vboulineau** [comment]
> This change will make us push the `latest` tag on `agent-dev` which previously was never tagged. There might be the expectation that `agent-dev:latest` points to the latest agent build, but this might not be the case since the job that regularly pushes to `agent-dev` won't update latest (only the manual jobs, when we trigger them). To avoid confusion, we might want to update the job that deploys to `agent-dev` so they also tag `latest` on every build.

Good catch, we should either not publish `latest` on `agent-dev` or use `master` branch only for `latest`

**Comment by KSerrania** [comment]
I want us to be able to publish `latest` on `agent-dev` (as I want the jobs to be strictly identical when they're run on `agent-dev` and `agent`), therefore I have the following proposal: the `deploy_latest` jobs are present on tagged deploy pipelines (ie. RCs and final artifacts), and on all other deploy pipelines on the condition that they're targeting master. That way, only images from `master` can end up as `agent-dev:latest`.

If we also want to ensure that `latest` is regularly updated, that'll require some additional work (eg. create new `dev_latest_*` jobs, or reuse the `dev_master_*` jobs in the `image_deploy` stage which push latest images on `agent-dev` on each master pipeline) that we can do in a later PR.

**Pull request merged**